src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java

```diff
@@ -52,94 +51,55 @@ public class RepositoryProgrammingExerciseParticipationResource extends Reposito
-        // Error case 5: The student can reset the repository only before and a tutor/instructor only after the due date has passed
-        if (repositoryAction == RepositoryActionType.RESET) {
-            boolean isOwner = true; // true for Solution- and TemplateProgrammingExerciseParticipation
-            if (participation instanceof StudentParticipation) {
-                isOwner = authCheckService.isOwnerOfParticipation((StudentParticipation) participation);
-            }
-            if (isStudent && programmingParticipation.isLocked()) {
-                // TODO: change to AccessForbiddenException
-                throw new IllegalAccessException();
-            }
-            // A tutor/instructor who is owner of the exercise should always be able to reset the repository
-            else if (!isStudent && !isOwner) {
-                // Check if a tutor is allowed to reset during the assessment
-                // Check for a regular course exercise
-                if (programmingExercise.isCourseExercise() && !programmingParticipation.isLocked()) {
-                    // TODO: change to AccessForbiddenException
-                    throw new IllegalAccessException();
-                }
-                // Check for an exam exercise, as it might not be locked but a student might still be allowed to submit
-                var optStudent = ((StudentParticipation) participation).getStudent();
-                if (optStudent.isPresent() && programmingExercise.isExamExercise()
-                        && examSubmissionService.isAllowedToSubmitDuringExam(programmingExercise, optStudent.get(), false)) {
-                    // TODO: change to AccessForbiddenException
-                    throw new IllegalAccessException();
-                }
-            }
+        try {
+            repositoryAccessService.checkAccessRepositoryElseThrow(programmingParticipation, userRepository.getUserWithGroupsAndAuthorities(), programmingExercise,
+                    repositoryActionType);
         }
-        // Error case 6: The user is not (any longer) allowed to submit to the exam/exercise. This check is only relevant for students.
-        // This must be a student participation as hasPermissions would have been false and an error already thrown
-        // But the student should still be able to access if they are notified for a related plagiarism case.
-        boolean isStudentParticipation = participation instanceof ProgrammingExerciseStudentParticipation;
-        if (isStudentParticipation && isStudent && !examSubmissionService.isAllowedToSubmitDuringExam(programmingExercise, user, false)
-                && !plagiarismService.wasUserNotifiedByInstructor(participationId, userRepository.getUser().getLogin())) {
-            // TODO: change to AccessForbiddenException
-            throw new IllegalAccessException();
+        catch (AccessUnauthorizedException | AccessForbiddenException e) {
```

Comment:
You can rethrow the AccessForbiddenException directly and only need to wrap the AccessUnauthorizedException.
```suggestion
        catch (AccessUnauthorizedException e) {
```